iceoryx_hoofs/source/posix_wrapper/shared_memory_object/shared_memory.cpp

```diff
@@ -33,73 +33,135 @@ namespace iox
+
+        if (m_openMode == OpenMode::PURGE_AND_CREATE)
+        {
+            IOX_DISCARD_RESULT(posixCall(iox_shm_unlink)(nameWithLeadingSlash.c_str())
+                                   .failureReturnValue(SharedMemory::INVALID_HANDLE)
+                                   .ignoreErrnos(ENOENT)
+                                   .evaluate());
+        }
+
+        auto result =
+            posixCall(iox_shm_open)(
+                nameWithLeadingSlash.c_str(),
+                getOflagsFor(m_accessMode,
+                             (m_openMode == OpenMode::OPEN_OR_CREATE) ? OpenMode::EXCLUSIVE_CREATE : m_openMode),
+                static_cast<mode_t>(m_filePermissions))
+                .failureReturnValue(SharedMemory::INVALID_HANDLE)
+                .suppressErrorMessagesForErrnos((m_openMode == OpenMode::OPEN_OR_CREATE) ? EEXIST : 0)
+                .evaluate();
+        if (result.has_error())
+        {
+            // if it was not possible to create the shm exclusively someone else has the
+            // ownership and we just try to open it
+            if (m_openMode == OpenMode::OPEN_OR_CREATE && result.get_error().errnum == EEXIST)
+            {
+                hasOwnership = false;
+                result = posixCall(iox_shm_open)(nameWithLeadingSlash.c_str(),
+                                                 getOflagsFor(m_accessMode, OpenMode::OPEN_EXISTING),
+                                                 static_cast<mode_t>(m_filePermissions))
+                             .failureReturnValue(SharedMemory::INVALID_HANDLE)
+                             .evaluate();
+                if (!result.has_error())
+                {
+                    sharedMemoryFileHandle = result->value;
+                    return cxx::success<SharedMemory>(SharedMemory(m_name, sharedMemoryFileHandle, hasOwnership));
+                }
+            }
+
+            printError();
+            return cxx::error<SharedMemoryError>(SharedMemory::errnoToEnum(result.get_error().errnum));
+        }
+        sharedMemoryFileHandle = result->value;
     }
 
-    if (!m_isInitialized)
+    if (hasOwnership)
     {
-        std::cerr << "Unable to create shared memory with the following properties [ name = " << name
-                  << ", access mode = " << ACCESS_MODE_STRING[static_cast<uint64_t>(accessMode)]
-                  << ", open mode = " << OPEN_MODE_STRING[static_cast<uint64_t>(openMode)]
-                  << ", mode = " << std::bitset<sizeof(mode_t)>(permissions) << ", sizeInBytes = " << size << " ]"
-                  << std::endl;
-        return;
+        auto result = posixCall(ftruncate)(sharedMemoryFileHandle, static_cast<int64_t>(m_size))
+                          .failureReturnValue(SharedMemory::INVALID_HANDLE)
+                          .evaluate();
+        if (result.has_error())
+        {
+            printError();
+            return cxx::error<SharedMemoryError>(SharedMemory::errnoToEnum(result->errnum));
```

Comment:
Shouldn't the file handle be closed before returning with an error?

Reply:
And we even should unlink the shared memory if we created it before. I added this.

Reply:
When we own the memory we even have to unlink it. I added this as well.